File: step1_start_project/src/main/java/migration/simple/controller/UserController.java
package migration.simple.controller;

import migration.simple.repository.UserRepository;
import migration.simple.responses.Response;
import migration.simple.responses.UserAddResponse;
import migration.simple.responses.UserResponse;
import migration.simple.types.User;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import java.util.Collections;
import java.util.List;
import java.util.Optional;


@RestController
public class UserController {
    private UserRepository userRepository;

    public UserController(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @GetMapping("users")
    public UserResponse users() {
        List<User> users = userRepository.findAllUsers();

        return new UserResponse(true, "return users", users);
    }

    @GetMapping("user/{id}")
    public UserResponse users(@PathVariable("id") Long userId) {
        Optional<User> user = userRepository.findUser(userId);

        return user
                .map(findUser -> new UserResponse(true, "find user with requested id", Collections.singletonList(findUser)))
                .orElseGet(() -> new UserResponse(false, "user not found", Collections.emptyList()));
    }

    @PutMapping(value = "user")
    public Response addUser(@RequestBody User user) {
        Optional<Long> addIndex = userRepository.addUser(user);

        return addIndex
                .map(index -> new UserAddResponse(true, "user add successfully", index))
                .orElseGet(() -> new UserAddResponse(false, "user not added", -1L));
    }

    @DeleteMapping("user/{id}")
    public Response deleteUser(@PathVariable("id") Long id) {
        boolean status = userRepository.deleteUser(id);

        if (status) {
            return new Response(true, "user has been deleted");
        } else {
            return new Response(false, "user not been deleted");
        }
    }
}

File: step4_migration_to_webflux/src/main/kotlin/migration/simple/controller/UserController.kt
package migration.simple.controller

import migration.simple.repository.UserRepository
import migration.simple.responses.DeleteResponse
import migration.simple.responses.UserAddResponse
import migration.simple.responses.UserResponse
import migration.simple.types.User
import org.springframework.web.reactive.function.server.RouterFunctionDsl
import org.springframework.web.reactive.function.server.ServerResponse.ok
import org.springframework.web.reactive.function.server.body
import reactor.core.publisher.Mono

open class UserController(private val userRepository: UserRepository) {
    fun nest(): RouterFunctionDsl.() -> Unit = {
        GET("/user") {
            ok().body(users())
        }
        GET("/user/{id}") {
            ok().body(user(it.pathVariable("id").toLong()))
        }
        PUT("/user") {
            ok().body(addUser(it.bodyToMono(User::class.java)))
        }
        DELETE("/user/{id}") {
            ok().body(deleteUser(it.pathVariable("id").toLong()))
        }
    }

    open fun users(): Mono<UserResponse> {
        val users = userRepository.findAllUsers()

        return Mono.just(UserResponse(true, "return users", users))
    }

    open fun user(userId: Long): Mono<UserResponse> {
        val user = userRepository.findUser(userId)

        val response = user
                ?.let { UserResponse(true, "find user with requested id", listOf(it)) }
                ?: UserResponse(false, "user not found", emptyList())

        return Mono.just(response)
    }

    open fun addUser(user: Mono<User>): Mono<UserAddResponse> = user.map {
        val addIndex = userRepository.addUser(it)

        addIndex
                ?.let { UserAddResponse(true, "user add successfully", it) }
                ?: UserAddResponse(false, "user not added", -1L)
    }

    open fun deleteUser(id: Long): Mono<DeleteResponse> {
        val status = userRepository.deleteUser(id)

        val result = if (status) {
            DeleteResponse(true, "user has been deleted")
        } else {
            DeleteResponse(false, "user not been deleted")
        }

        return Mono.just(result)
    }
}
